Set UserAgent and network settings for the msbuild restore task

emgarten opened the pull request:
Set UserAgent and network settings for the msbuild restore task

Copying XPlat network settings over to the MSBuild RestoreTask. This includes:
* UserAgent
* Allowed SSL protocols
* Connection limit

Fixes https://github.com/NuGet/Home/issues/3905

//cc @joelverhagen @mishra14 @alpaix @drewgil @dtivel @jainaashish @zhili1208

mishra14 commented on src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/XPlatUtility.cs:
```diff
@@ -44,21 +44,5 @@ public static void SetUserAgent()
             UserAgent.SetUserAgentString(new UserAgentStringBuilder("NuGet xplat"));
```
Please add Copyright and Licensing information to this file.

emgarten replied:
@mishra14 please open an issue to track automating this process as part of build.ps1/build.sh to avoid noise in PRs. Thanks.

mishra14 replied:
https://github.com/NuGet/Home/issues/3946

mishra14 commented on src/NuGet.Core/NuGet.Build.Tasks/BuildTasksUtility.cs:
```diff
@@ -1,7 +1,10 @@
 ﻿using System.Collections.Generic;
```
Please add Copyright and Licensing information to this file.

alpaix commented on src/NuGet.Core/NuGet.Protocol.Core.v3/Utility/ProtocolUtility.cs:
```diff
@@ -0,0 +1,30 @@
+        /// <summary>
+        /// Set ServicePointManager.DefaultConnectionLimit
+        /// </summary>
+        public static void SetConnectionLimit()
```
Can this be a method of `NetworkProtocolUtility` instead?

emgarten replied:
Good idea, I'll move it over there.

mishra14 commented on src/NuGet.Core/NuGet.Protocol.Core.v3/Utility/ProtocolUtility.cs:
```diff
@@ -0,0 +1,30 @@
+        /// </summary>
+        public static void SetConnectionLimit()
+        {
+#if !IS_CORECLR
```
does this not apply to `CORECLR` ?

dtivel commented on src/NuGet.Core/NuGet.Build.Tasks/BuildTasksUtility.cs:
```diff
@@ -1,7 +1,10 @@
 using Microsoft.Build.Framework;
+using NuGet.Common;
 using NuGet.ProjectModel;
+using NuGet.Protocol.Core.Types;
```
The changes to this file appear to be unnecessary.

dtivel commented on src/NuGet.Core/NuGet.Protocol.Core.v3/Utility/ProtocolUtility.cs:
```diff
@@ -0,0 +1,30 @@
+        /// <summary>
+        /// Set ServicePointManager.DefaultConnectionLimit
+        /// </summary>
+        public static void SetConnectionLimit()
```
Can [NuGet.CommandLine.Program.MainCore(...)](https://github.com/NuGet/NuGet.Client/blob/4cccb13833ad29d6a0bcff055460d964f1b49cfe/src/NuGet.Clients/NuGet.CommandLine/Program.cs#L70) be updated to use this?

dtivel commented on src/NuGet.Core/NuGet.Protocol.Core.v3/Utility/ProtocolUtility.cs:
```diff
@@ -0,0 +1,30 @@
+
+namespace NuGet.Protocol
+{
+    public class ProtocolUtility
```
`static`?

emgarten replied:
Good catch, it should have been. This will get moved over to NetworkProtocolUtility

dtivel commented on src/NuGet.Core/NuGet.Build.Tasks/RestoreTask.cs:
```diff
@@ -192,5 +207,18 @@ private static MSBuildTaskItem GetMSBuildItem(ITaskItem item)
+
+#if IS_CORECLR
+            UserAgent.SetUserAgentString(new UserAgentStringBuilder(agent)
+                .WithOSDescription(RuntimeInformation.OSDescription));
```
[`RuntimeInformation`](https://docs.microsoft.com/en-us/dotnet/core/api/system.runtime.interopservices.runtimeinformation) is in the `System.Runtime.InteropServices` namespace, and yet I do not see a `using` directive for it.  I expect a compilation error here when `IS_CORECLR` is defined.

emgarten replied:
The CI agrees :smile:

emgarten committed:
Set UserAgent and network settings for the msbuild restore task

Copying XPlat network settings over to the MSBuild RestoreTask. This includes:
* UserAgent
* Allowed SSL protocols
* Connection limit

Fixes https://github.com/NuGet/Home/issues/3905
